tests/phpunit/tests/rest-api/rest-block-type-controller.php

```diff
@@ -301,6 +306,46 @@ public function test_get_item_defaults() {
 		$this->assertNull( $data['example'] );
 		$this->assertNull( $data['textdomain'] );
 		$this->assertFalse( $data['is_dynamic'] );
+		$this->assertSameSets( array(), $data['variations'] );
+	}
+
+	public function test_get_variation() {
+		$block_type = 'fake/variations';
+		$settings   = array(
+			'title'       => 'variations block test',
+			'description' => 'a variations block test',
+			'attributes'  => array( 'kind' => array( 'type' => 'string' ) ),
+			'variations'  => array(
+				array(
+					'name'       => 'Foo',
+					'title'      => 'Foo Variation',
+					'attributes' => array( 'kind' => 'foo' ),
+				),
+			),
+		);
+		register_block_type( $block_type, $settings );
+		wp_set_current_user( self::$admin_id );
+		$request  = new WP_REST_Request( 'GET', '/wp/v2/block-types/' . $block_type );
+		$response = rest_get_server()->dispatch( $request );
+		$data     = $response->get_data();
+		$this->assertSame( $block_type, $data['name'] );
+		$this->assertArrayHasKey( 'variations', $data );
+		$this->assertSame( 1, count( $data['variations'] ) );
+		$variation = $data['variations'][0];
+		$this->assertArrayHasKey( 'attributes', $variation );
+		//TODO: this is getting wiped by:
+		// $data[ $extra_field ] = rest_sanitize_value_from_schema( $field, $schema['properties'][ $extra_field ] );
+		// in class-wp-rest-block-types-controller.php
```

Comment:
🤔 I'm thinking through this one, but I could use some advice on how to handle this. Schema validation currently wipes variation attribute values. Top level block attributes don't use the schema and instead have special handling.

<img width="854" alt="Screen Shot 2021-02-23 at 12 30 45 PM" src="https://user-images.githubusercontent.com/1270189/108915652-0ca30100-75e2-11eb-92cd-a103143eb503.png">

We can see this by testing on https://github.com/WordPress/gutenberg/pull/29095 in console
```
wp.apiFetch( { path: '/wp/v2/block-types' } ).then( ( data ) => { 
    const nav = data.filter( ( blockType ) => { 
        return blockType.name === 'core/navigation-link';
    } );
    console.log( nav.pop().variations );
} );
```

If we have a WP API REST client for testing, let me know ;)

Reply:
Folks have permission to commit directly, if y'all happen to spot something.

Reply:
Why don't they use the schema?

Reply:
Here's what the top level `attributes` calls. I might be misreading it, but I don't think it uses the schema definition.
https://github.com/WordPress/wordpress-develop/blob/50506a8e6f54e1b36ac8ab2476600a4e48a79086/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L250

Variations is currently calling:
https://github.com/WordPress/wordpress-develop/blob/50506a8e6f54e1b36ac8ab2476600a4e48a79086/src/wp-includes/rest-api/endpoints/class-wp-rest-block-types-controller.php#L288

So for this example:
```php
...
'variations'  => array(
      array(
	      'name'       => 'Foo',
	      'title'      => 'Foo Variation',
	      'attributes' => array( 'kind' => 'foo' ),
      ),
   ),
);
```

It gets validated to:
```php
...
'variations'  => array(
      array(
	      'name'       => 'Foo',
	      'title'      => 'Foo Variation',
              //string foo is lost
	      'attributes' => array( 'kind' => array() ), 
      ),
   ),
);
```

I've probably misdefined the schema here, but I'm wondering how useful the variations attributes is. For the caller I'd expect that we'd validate by the block attributes schema, similar to what we're doing in $block_type->prepare_attributes_for_render.

Eg, it'd throw out values that are not defined as block attributes. So let's say we have a block with a 'type' attribute. If I provide a variation with a 'type' and 'foo' attribute I'd expect it to remove the 'foo' attribute on the variation.

Reply:
Fixed this one by simplifying the attributes schema in https://github.com/WordPress/wordpress-develop/pull/1015/commits/496c72edb5ad2d68d5348e7ebcf793ee2d7669fc. We can maybe punt stricter validation for the future. Eg this literally just checks that we're passing an object vs if the attribute is part of the block attribute definition.

<img width="747" alt="Screen Shot 2021-02-24 at 2 23 33 PM" src="https://user-images.githubusercontent.com/1270189/109074494-ae8f2000-76ac-11eb-9d55-f4a597d3041f.png">

Reply:
Yeah this is fine.